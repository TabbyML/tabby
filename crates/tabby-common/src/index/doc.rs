pub mod fields {
    pub const TITLE: &str = "title";
    pub const LINK: &str = "link";
    pub const CHUNK_TEXT: &str = "chunk_text";
}
