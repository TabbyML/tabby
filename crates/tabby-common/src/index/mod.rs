pub mod code;
pub mod doc;

use std::borrow::Cow;

use lazy_static::lazy_static;
use tantivy::{
    query::{BooleanQuery, ConstScoreQuery, ExistsQuery, Occur, Query, TermQuery},
    schema::{
        Field, IndexRecordOption, JsonObjectOptions, Schema, TextFieldIndexing, FAST, INDEXED,
        STORED, STRING,
    },
    Term,
};

pub struct IndexSchema {
    pub schema: Schema,

    // === Fields for both document and chunk ===
    /// Corpus for the document, each corpus comes with its own schema for json fields (field_attributes / field_chunk_attributes)
    /// See ./doc or ./code as an example
    pub field_corpus: Field,

    /// Unique identifier (within corpus) for a group of documents.
    pub field_source_id: Field,

    /// Unique identifier (within corpus) for the document, each document could have multiple chunks indexed.
    pub field_id: Field,

    /// Last updated time for the document in index.
    pub field_updated_at: Field,
    // ==========================================

    // === Fields for document ===
    /// JSON attributes for the document, it's only stored but not indexed.
    pub field_attributes: Field,
    // ===========================

    // === Fields for chunk ===
    pub field_chunk_id: Field,
    /// JSON attributes for the chunk, it's indexed (thus can be used as filter in query) and stored.
    pub field_chunk_attributes: Field,
    /// Matching tokens for the chunk, it's indexed but not stored..
    pub field_chunk_tokens: Field,
    // =========================
}

const FIELD_CHUNK_ID: &str = "chunk_id";

pub mod corpus {
    pub const CODE: &str = "code";
    pub const WEB: &str = "web";
}

impl IndexSchema {
    pub fn instance() -> &'static Self {
        &INDEX_SCHEMA
    }

    fn new() -> Self {
        let mut builder = Schema::builder();

        let field_corpus = builder.add_text_field("corpus", STRING | FAST);
        let field_source_id = builder.add_text_field("source_id", STRING | FAST);
        let field_id = builder.add_text_field("id", STRING | STORED);

        let field_updated_at = builder.add_date_field("updated_at", INDEXED);
        let field_attributes = builder.add_text_field("attributes", STORED);

        let field_chunk_id = builder.add_text_field(FIELD_CHUNK_ID, STRING | FAST | STORED);
        let field_chunk_attributes = builder.add_json_field(
            "chunk_attributes",
            JsonObjectOptions::default()
                .set_stored()
                .set_indexing_options(
                    TextFieldIndexing::default()
                        .set_tokenizer("raw")
                        .set_fieldnorms(true)
                        .set_index_option(tantivy::schema::IndexRecordOption::Basic)
                        .set_fieldnorms(true),
                ),
        );

        let field_chunk_tokens = builder.add_text_field("chunk_tokens", STRING);
        let schema = builder.build();

        Self {
            schema,
            field_id,
            field_source_id,
            field_corpus,
            field_updated_at,
            field_attributes,

            field_chunk_id,
            field_chunk_attributes,
            field_chunk_tokens,
        }
    }

    pub fn source_query(&self, corpus: &str, source_id: &str) -> impl Query {
        let source_id_query = TermQuery::new(
            Term::from_field_text(self.field_source_id, source_id),
            tantivy::schema::IndexRecordOption::Basic,
        );

        BooleanQuery::new(vec![
            // Must match the corpus
            (Occur::Must, self.corpus_query(corpus)),
            // Must match the source id
            (Occur::Must, Box::new(source_id_query)),
        ])
    }

    /// Build a query to find the document with the given `doc_id`.
    pub fn doc_query(&self, corpus: &str, doc_id: &str) -> impl Query {
        let doc_id_query = TermQuery::new(
            Term::from_field_text(self.field_id, doc_id),
            tantivy::schema::IndexRecordOption::Basic,
        );

        BooleanQuery::new(vec![
            // Must match the corpus
            (Occur::Must, self.corpus_query(corpus)),
            // Must match the doc id
            (Occur::Must, Box::new(doc_id_query)),
            // Exclude chunk documents
            (
                Occur::MustNot,
                Box::new(ExistsQuery::new_exists_query(FIELD_CHUNK_ID.into())),
            ),
        ])
    }

    /// Build a query to find the document with the given `doc_id`, include chunks.
    pub fn doc_query_with_chunks(&self, corpus: &str, doc_id: &str) -> impl Query {
        let doc_id_query = TermQuery::new(
            Term::from_field_text(self.field_id, doc_id),
            tantivy::schema::IndexRecordOption::Basic,
        );

        BooleanQuery::new(vec![
            // Must match the corpus
            (Occur::Must, self.corpus_query(corpus)),
            // Must match the doc id
            (Occur::Must, Box::new(doc_id_query)),
        ])
    }

    pub fn corpus_query(&self, corpus: &str) -> Box<dyn Query> {
        Box::new(TermQuery::new(
            Term::from_field_text(self.field_corpus, corpus),
            tantivy::schema::IndexRecordOption::Basic,
        ))
    }
}

lazy_static! {
    static ref INDEX_SCHEMA: IndexSchema = IndexSchema::new();
}

pub fn binarize_embedding<'a>(
    embedding: impl Iterator<Item = &'a f32> + 'a,
) -> impl Iterator<Item = String> + 'a {
    embedding.enumerate().map(|(i, value)| {
        if *value <= 0.0 {
            format!("embedding_zero_{}", i)
        } else {
            format!("embedding_one_{}", i)
        }
    })
}

pub fn embedding_tokens_query<'a>(
    embedding_dims: usize,
    embedding: impl Iterator<Item = &'a f32> + 'a,
) -> BooleanQuery {
    let schema = IndexSchema::instance();
    let iter = binarize_embedding(embedding).map(Cow::Owned);
    new_multiterms_const_query(schema.field_chunk_tokens, embedding_dims, iter)
}

fn new_multiterms_const_query<'a>(
    field: Field,
    embedding_dims: usize,
    terms: impl Iterator<Item = Cow<'a, str>> + 'a,
) -> BooleanQuery {
    let subqueries: Vec<Box<dyn Query>> = terms
        .map(|text| {
            let term = Term::from_field_text(field, text.as_ref());
            let term_query: Box<dyn Query> =
                Box::new(TermQuery::new(term, IndexRecordOption::Basic));

            let score = 1.0 / embedding_dims as f32;
            let boxed: Box<dyn Query> = Box::new(ConstScoreQuery::new(term_query, score));

            boxed
        })
        .collect();

    BooleanQuery::union(subqueries)
}

#[cfg(test)]
mod tests {

    use tantivy::{
        collector::TopDocs,
        query::Query,
        schema::{Schema, STRING},
        Index, IndexWriter, TantivyDocument,
    };

    use super::*;

    #[test]
    fn test_new_multiterms_const_query() -> anyhow::Result<()> {
        let mut schema_builder = Schema::builder();
        let field1 = schema_builder.add_text_field("field1", STRING);
        let schema = schema_builder.build();
        let index = Index::create_in_ram(schema);
        {
            let mut index_writer: IndexWriter = index.writer(15_000_000)?;

            // doc1
            let mut doc = TantivyDocument::new();
            doc.add_text(field1, "value1");
            doc.add_text(field1, "value2");
            doc.add_text(field1, "value3");
            index_writer.add_document(doc)?;

            // doc2
            let mut doc = TantivyDocument::new();
            doc.add_text(field1, "value2");
            doc.add_text(field1, "value4");
            index_writer.add_document(doc)?;

            index_writer.commit()?;
        }
        let reader = index.reader()?;
        let searcher = reader.searcher();

        {
            let query = new_multiterms_const_query(
                field1,
                4,
                vec!["value1", "value3"].into_iter().map(Cow::Borrowed),
            );

            let top_docs = searcher.search(&query, &TopDocs::with_limit(1))?;
            eprintln!("explain {:?}", query.explain(&searcher, top_docs[0].1)?);

            assert_eq!(top_docs.len(), 1, "Expected 1 document");
            assert_eq!(top_docs[0].0, 0.5);
        }

        {
            let query = new_multiterms_const_query(
                field1,
                4,
                vec!["value1", "value2", "value3"]
                    .into_iter()
                    .map(Cow::Borrowed),
            );

            let top_docs = searcher.search(&query, &TopDocs::with_limit(1))?;

            assert_eq!(top_docs.len(), 1, "Expected 1 document");
            assert_eq!(top_docs[0].0, 0.75);
        }

        Ok(())
    }
}
