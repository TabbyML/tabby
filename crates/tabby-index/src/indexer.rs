use std::collections::HashSet;

use anyhow::{bail, Result};
use async_stream::stream;
use futures::{stream::BoxStream, Stream, StreamExt};
use serde_json::json;
use tabby_common::{
    index::{IndexSchema, FIELD_SOURCE_ID},
    path,
};
use tantivy::{
    aggregation::{
        agg_req::Aggregation,
        agg_result::{AggregationResult, BucketResult},
        AggregationCollector, Key,
    },
    collector::TopDocs,
    doc,
    query::AllQuery,
    schema::{self, Value},
    DocAddress, DocSet, IndexWriter, Searcher, TantivyDocument, Term, TERMINATED,
};
use tokio::{sync::mpsc, task::JoinHandle};
use tracing::{debug, warn};

use crate::tantivy_utils::open_or_create_index;

pub struct IndexId {
    pub source_id: String,
    pub id: String,
}

pub trait ToIndexId {
    fn to_index_id(&self) -> IndexId;
}

#[async_trait::async_trait]
pub trait IndexAttributeBuilder<T>: Send + Sync {
    /// Build document level attributes, these attributes are only stored but not indexed.
    async fn build_attributes(&self, document: &T) -> serde_json::Value;

    /// Build chunk level attributes, these attributes are stored and indexed.
    async fn build_chunk_attributes<'a>(
        &self,
        document: &'a T,
    ) -> BoxStream<'a, JoinHandle<Result<(Vec<String>, serde_json::Value)>>>;
}

pub struct TantivyDocBuilder<T> {
    corpus: &'static str,
    builder: Box<dyn IndexAttributeBuilder<T>>,
}

impl<T: ToIndexId> TantivyDocBuilder<T> {
    pub fn new(corpus: &'static str, builder: impl IndexAttributeBuilder<T> + 'static) -> Self {
        Self {
            corpus,
            builder: Box::new(builder),
        }
    }

    pub async fn build(
        &self,
        document: T,
    ) -> (
        String,
        impl Stream<Item = JoinHandle<Option<TantivyDocument>>> + '_,
    ) {
        let schema = IndexSchema::instance();
        let IndexId { source_id, id } = document.to_index_id();

        let now = tantivy::time::OffsetDateTime::now_utc();
        let updated_at = tantivy::DateTime::from_utc(now);

        let cloned_id = id.clone();
        let doc_id = id.clone();
        let doc_attributes = self.builder.build_attributes(&document).await;
        let s = stream! {
            let (tx, mut rx) = mpsc::channel(32);

            for await chunk_doc in self.build_chunks(cloned_id, source_id.clone(), updated_at, document).await {
                let tx = tx.clone();
                let doc_id = doc_id.clone();
                yield tokio::spawn(async move {
                    match chunk_doc.await {
                        Ok(Ok(doc)) => {
                            Some(doc)
                        }
                        Ok(Err(e)) => {
                            warn!("Failed to build chunk for document '{}': {}", doc_id, e);
                            tx.send(()).await.unwrap_or_else(|e| {
                                warn!("Failed to send error signal for document '{}': {}", doc_id, e);
                            });
                            None
                        }
                        Err(e) => {
                            warn!("Failed to call build chunk '{}': {}", doc_id, e);
                            tx.send(()).await.unwrap_or_else(|e| {
                                warn!("Failed to send error signal for document '{}': {}", doc_id, e);
                            });
                            None
                        }
                    }
                });
            };

            // drop tx to signal the end of the stream
            // the cloned is dropped in its own thread
            drop(tx);

            let mut doc = doc! {
                schema.field_id => doc_id,
                schema.field_source_id => source_id,
                schema.field_corpus => self.corpus,
                schema.field_attributes => doc_attributes,
                schema.field_updated_at => updated_at,
            };

            yield tokio::spawn(async move {
                let mut failed_count = 0;
                while let Some(_) = rx.recv().await {
                    failed_count += 1;
                }
                if failed_count > 0 {
                    doc.add_u64(schema.field_failed_chunks_count, failed_count as u64);
                }
                Some(doc)
             });
        };

        (id, s)
    }

    async fn build_chunks(
        &self,
        id: String,
        source_id: String,
        updated_at: tantivy::DateTime,
        document: T,
    ) -> impl Stream<Item = JoinHandle<Result<TantivyDocument>>> + '_ {
        let kind = self.corpus;
        stream! {
            let schema = IndexSchema::instance();
            for await (chunk_id, task) in self.builder.build_chunk_attributes(&document).await.enumerate() {
                let id = id.clone();
                let source_id = source_id.clone();

                yield tokio::spawn(async move {
                    let built_chunk_attributes_result = task.await?;
                    let (tokens, chunk_attributes) = built_chunk_attributes_result?;

                    let mut doc = doc! {
                        schema.field_id => id,
                        schema.field_source_id => source_id,
                        schema.field_corpus => kind,
                        schema.field_updated_at => updated_at,
                        schema.field_chunk_id => format!("{}-{}", id, chunk_id),
                        schema.field_chunk_attributes => chunk_attributes,
                    };

                    for token in &tokens {
                        doc.add_text(schema.field_chunk_tokens, token);
                    }

                    Ok(doc)
                });
            }
        }
    }

    pub async fn backfill_doc_attributes(
        &self,
        origin: &TantivyDocument,
        doc: &T,
    ) -> TantivyDocument {
        let schema = IndexSchema::instance();
        let mut doc = doc! {
            schema.field_id => get_text(origin, schema.field_id),
            schema.field_source_id => get_text(origin, schema.field_source_id).to_string(),
            schema.field_corpus => get_text(origin, schema.field_corpus).to_string(),
            schema.field_attributes => self.builder.build_attributes(doc).await,
            schema.field_updated_at => get_date(origin, schema.field_updated_at),
        };
        if let Some(failed_chunks) = get_number_optional(origin, schema.field_failed_chunks_count) {
            doc.add_u64(schema.field_failed_chunks_count, failed_chunks as u64);
        }

        doc
    }
}

pub struct Indexer {
    corpus: String,
    searcher: Searcher,
    writer: IndexWriter,
}

impl Indexer {
    pub fn new(corpus: &str) -> Self {
        let doc = IndexSchema::instance();
        let (_, index) = open_or_create_index(&doc.schema, &path::index_dir());
        let writer = index
            .writer(150_000_000)
            .expect("Failed to create index writer");
        let reader = index.reader().expect("Failed to create index reader");

        Self {
            corpus: corpus.to_owned(),
            searcher: reader.searcher(),
            writer,
        }
    }

    pub async fn add(&self, document: TantivyDocument) {
        self.writer
            .add_document(document)
            .expect("Failed to add document");
    }

    pub async fn get_doc(&self, id: &str) -> Result<TantivyDocument> {
        let schema = IndexSchema::instance();
        let query = schema.doc_query(&self.corpus, id);
        let docs = match self.searcher.search(&query, &TopDocs::with_limit(1)) {
            Ok(docs) => docs,
            Err(e) => {
                debug!("query tantivy error: {}", e);
                return Err(e.into());
            }
        };
        if docs.is_empty() {
            bail!("Document not found: {}", id);
        }

        self.searcher
            .doc(docs.first().unwrap().1)
            .map_err(|e| e.into())
    }

    pub fn delete(&self, id: &str) {
        let schema = IndexSchema::instance();
        let _ = self
            .writer
            .delete_query(Box::new(schema.doc_query_with_chunks(&self.corpus, id)));
    }

    pub fn delete_doc(&self, id: &str) {
        let schema = IndexSchema::instance();
        let _ = self
            .writer
            .delete_query(Box::new(schema.doc_query(&self.corpus, id)));
    }

    pub fn commit(mut self) {
        self.writer.commit().expect("Failed to commit changes");
        self.writer
            .wait_merging_threads()
            .expect("Failed to wait for merging threads");
    }

    // Check whether the document ID presents in the corpus.
    pub fn is_indexed(&self, id: &str) -> bool {
        let schema = IndexSchema::instance();
        let query = schema.doc_query(&self.corpus, id);
        let Ok(docs) = self.searcher.search(&query, &TopDocs::with_limit(1)) else {
            return false;
        };
        !docs.is_empty()
    }

    /// Iterates over all the document IDs in the corpus.
    pub fn iter_ids(&self) -> impl Stream<Item = String> + '_ {
        let schema = IndexSchema::instance();

        stream! {
            // Based on https://github.com/quickwit-oss/tantivy/blob/main/examples/iterating_docs_and_positions.rs
            for (segment_ordinal, segment_reader) in self.searcher.segment_readers().iter().enumerate() {
                let Ok(inverted_index) = segment_reader.inverted_index(schema.field_corpus) else {
                    continue;
                };

                let term_corpus = Term::from_field_text(schema.field_corpus, &self.corpus);
                let Ok(Some(mut postings)) = inverted_index.read_postings(&term_corpus, tantivy::schema::IndexRecordOption::Basic) else {
                    continue;
                };

                let mut doc_id = postings.doc();
                while doc_id != TERMINATED {
                    if !segment_reader.is_deleted(doc_id) {
                        let doc_address = DocAddress::new(segment_ordinal as u32, doc_id);
                        let doc: TantivyDocument = self.searcher.doc(doc_address).expect("Failed to read document");

                        // Skip chunks, as we only want to iterate over the main docs
                        if doc.get_first(schema.field_chunk_id).is_none() {
                            let id = get_text(&doc, schema.field_id);
                            yield id.to_owned();
                        }
                    }
                    doc_id = postings.advance();
                }
            }
        }
    }

    pub fn is_indexed_after(&self, id: &str, time: chrono::DateTime<chrono::Utc>) -> bool {
        let schema = IndexSchema::instance();
        let query = schema.doc_indexed_after(&self.corpus, id, time);
        let Ok(docs) = self.searcher.search(&query, &TopDocs::with_limit(1)) else {
            return false;
        };

        !docs.is_empty()
    }

    /// Check whether the document has failed chunks.
    ///
    /// failed chunks tracks the number of embedding indexing failed chunks for a document.
    pub fn has_failed_chunks(&self, id: &str) -> bool {
        let schema = IndexSchema::instance();
        let query = schema.doc_has_failed_chunks(&self.corpus, id);
        let Ok(docs) = self.searcher.search(&query, &TopDocs::with_limit(1)) else {
            return false;
        };

        !docs.is_empty()
    }

    // Check whether the document has attribute field.
    pub fn has_attribute_field(&self, id: &str, field: &str) -> bool {
        let schema = IndexSchema::instance();
        let query = schema.doc_has_attribute_field(&self.corpus, id, field);
        match self.searcher.search(&query, &TopDocs::with_limit(1)) {
            Ok(docs) => !docs.is_empty(),
            Err(e) => {
                debug!("query tantivy error: {}", e);
                false
            }
        }
    }
}

pub struct IndexGarbageCollector {
    searcher: Searcher,
    writer: IndexWriter,
}

impl IndexGarbageCollector {
    pub fn new() -> Self {
        let doc = IndexSchema::instance();
        let (_, index) = open_or_create_index(&doc.schema, &path::index_dir());
        let writer = index
            .writer(150_000_000)
            .expect("Failed to create index writer");
        let reader = index.reader().expect("Failed to create index reader");

        Self {
            searcher: reader.searcher(),
            writer,
        }
    }

    pub fn garbage_collect(&self, active_source_ids: &[String]) -> anyhow::Result<()> {
        let source_ids: HashSet<_> = active_source_ids.iter().collect();

        let count_aggregation: Aggregation = serde_json::from_value(json!({
            "terms": {
                "field": FIELD_SOURCE_ID,
            }
        }))
        .unwrap();

        let collector = AggregationCollector::from_aggs(
            vec![("count".to_owned(), count_aggregation)]
                .into_iter()
                .collect(),
            Default::default(),
        );

        let res = self.searcher.search(&AllQuery, &collector)?;
        let Some(AggregationResult::BucketResult(BucketResult::Terms { buckets, .. })) =
            res.0.get("count")
        else {
            bail!("Failed to get source_id count");
        };
        for source_id in buckets {
            let count = source_id.doc_count;
            let Key::Str(source_id) = &source_id.key else {
                warn!("Failed to get source_id key as string");
                continue;
            };

            if !source_ids.contains(source_id) {
                debug!("Deleting {} documents for source_id: {}", count, source_id,);
                self.delete_by_source_id(source_id);
            }
        }

        Ok(())
    }

    fn delete_by_source_id(&self, source_id: &str) {
        let schema = IndexSchema::instance();
        let _ = self
            .writer
            .delete_query(Box::new(schema.source_id_query(source_id)));
    }

    pub fn commit(mut self) {
        self.writer.commit().expect("Failed to commit changes");
        self.writer
            .wait_merging_threads()
            .expect("Failed to wait for merging threads");
    }
}

fn get_text(doc: &TantivyDocument, field: schema::Field) -> &str {
    doc.get_first(field).unwrap().as_str().unwrap()
}

fn get_date(doc: &TantivyDocument, field: schema::Field) -> tantivy::DateTime {
    doc.get_first(field).unwrap().as_datetime().unwrap()
}

fn get_number_optional(doc: &TantivyDocument, field: schema::Field) -> Option<i64> {
    doc.get_first(field)?.as_i64()
}
