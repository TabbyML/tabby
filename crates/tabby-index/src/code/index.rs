use std::{path::Path, pin::pin, sync::Arc};

use anyhow::Result;
use async_stream::stream;
use futures::StreamExt;
use ignore::{DirEntry, Walk};
use tabby_common::index::{code, corpus};
use tabby_inference::Embedding;
use tracing::warn;

use super::{
    create_code_builder,
    intelligence::{CodeIntelligence, SourceCode},
    CodeRepository,
};
use crate::indexer::{Indexer, TantivyDocBuilder};

// Magic numbers
static MAX_LINE_LENGTH_THRESHOLD: usize = 300;
static AVG_LINE_LENGTH_THRESHOLD: f32 = 150f32;
static MIN_ALPHA_NUM_FRACTION: f32 = 0.25f32;
static MAX_NUMBER_OF_LINES: usize = 100000;
static MAX_NUMBER_FRACTION: f32 = 0.5f32;

pub async fn index_repository(
    embedding: Arc<dyn Embedding>,
    repository: &CodeRepository,
    commit: &str,
) {
    let total_files = Walk::new(repository.dir()).count();
    let file_stream = stream! {
        for file in Walk::new(repository.dir()) {
            let file = match file {
                Ok(file) => file,
                Err(e) => {
                    warn!("Failed to walk file tree for indexing: {e}");
                    continue;
                }
            };

            yield file;
        }
    }
    // Commit every 100 files
    .chunks(100);

    let mut file_stream = pin!(file_stream);

    let mut count_files = 0;
    let mut count_chunks = 0;
    while let Some(files) = file_stream.next().await {
        count_files += files.len();
        count_chunks += add_changed_documents(repository, commit, embedding.clone(), files).await;
        logkit::info!("Processed {count_files}/{total_files} files, updated {count_chunks} chunks",);
    }
}

pub async fn garbage_collection() {
    let index = Indexer::new(corpus::CODE);
    stream! {
        let mut num_to_keep = 0;
        let mut num_to_delete = 0;

        for await id in index.iter_ids() {
            let Some(source_file_id) = SourceCode::source_file_id_from_id(&id) else {
                warn!("Failed to extract source file id from index id: {id}");
                num_to_delete += 1;
                index.delete(&id);
                continue;
            };

            if CodeIntelligence::check_source_file_id_matched(source_file_id) {
                num_to_keep += 1;
            } else {
                num_to_delete += 1;
                index.delete(&id);
            }
        }

        logkit::info!("Finished garbage collection for code index: {num_to_keep} items kept, {num_to_delete} items removed");
        index.commit();
    }.collect::<()>().await;
}

async fn add_changed_documents(
    repository: &CodeRepository,
    commit: &str,
    embedding: Arc<dyn Embedding>,
    files: Vec<DirEntry>,
) -> usize {
    let concurrency = std::cmp::max(std::thread::available_parallelism().unwrap().get() * 4, 64);
    let builder = Arc::new(create_code_builder(Some(embedding)));
    let index = Arc::new(Indexer::new(corpus::CODE));
    let cloned_index = index.clone();

    let mut count_docs = 0;
    stream! {
        for file in files {
            let Some(key) = CodeIntelligence::compute_source_file_id(file.path()) else {
                continue;
            };

            let id = SourceCode::to_index_id(&repository.source_id, &key).id;

            // Skip if already indexed and has no failed chunks,
            // when skip, we should check if the document needs to be backfilled.
            if !require_updates(cloned_index.clone(), &id) {
                backfill_commit_in_doc_if_needed(
                    builder.clone(),
                    cloned_index.clone(),
                    &id,
                    repository,
                    commit,
                    file.path()).await.unwrap_or_else(|e| {
                        warn!("Failed to backfill commit for {id}: {e}");
                    }
                );
                continue;
            }

            let Some(code) = CodeIntelligence::compute_source_file(repository, commit, file.path()) else {
                continue;
            };

            if !is_valid_file(&code) {
                continue;
            }

            let (_, s) = builder.build(code).await;
            // must delete before adding, otherwise the some fields like failed_chunks_count will remain
            cloned_index.delete(&id);
            for await task in s {
                yield task;
            }
        }
    }
    .buffer_unordered(concurrency)
    .filter_map(|x| async { x.ok().flatten() })
    .for_each(|x| {
        count_docs += 1;
        index.add(x)
    })
    .await;

    match Arc::try_unwrap(index) {
        Ok(index) => index.commit(),
        Err(_) => {
            panic!("Failed to commit code index");
        }
    };

    count_docs
}

fn require_updates(indexer: Arc<Indexer>, id: &str) -> bool {
    if indexer.is_indexed(id) && !indexer.has_failed_chunks(id) {
        return false;
    };

    true
}

// v0.23.0 add the commit field to the code document.
async fn backfill_commit_in_doc_if_needed(
    builder: Arc<TantivyDocBuilder<SourceCode>>,
    indexer: Arc<Indexer>,
    id: &str,
    repository: &CodeRepository,
    commit: &str,
    path: &Path,
) -> Result<()> {
    if indexer.has_attribute_field(id, code::fields::ATTRIBUTE_COMMIT) {
        return Ok(());
    }

    let code = CodeIntelligence::compute_source_file(repository, commit, path)
        .ok_or_else(|| anyhow::anyhow!("Failed to compute source file"))?;
    if !is_valid_file(&code) {
        anyhow::bail!("Invalid file");
    }

    let origin = indexer.get_doc(id).await?;
    indexer.delete_doc(id);
    indexer
        .add(builder.backfill_doc_attributes(&origin, &code).await)
        .await;

    Ok(())
}

fn is_valid_file(file: &SourceCode) -> bool {
    file.max_line_length <= MAX_LINE_LENGTH_THRESHOLD
        && file.avg_line_length <= AVG_LINE_LENGTH_THRESHOLD
        && file.alphanum_fraction >= MIN_ALPHA_NUM_FRACTION
        && file.num_lines <= MAX_NUMBER_OF_LINES
        && file.number_fraction <= MAX_NUMBER_FRACTION
}

#[cfg(test)]
mod tests {
    use futures::StreamExt;
    use insta::assert_snapshot;

    use crate::code::intelligence::CodeIntelligence;

    #[tokio::test]
    async fn test_code_splitter() {
        // First file, tabby-inference/src/decoding.rs
        let file_contents = include_str!("../../../tabby-inference/src/decoding.rs");

        let rust_chunks = CodeIntelligence::chunks(file_contents, "rust")
            .map(|(_, chunk)| chunk)
            .collect::<Vec<_>>()
            .await;

        assert_snapshot!(format!("{:#?}", rust_chunks));

        let text_chunks = CodeIntelligence::chunks(file_contents, "unknown")
            .map(|(_, chunk)| chunk)
            .collect::<Vec<_>>()
            .await;

        assert_snapshot!(format!("{:#?}", text_chunks));

        // Second file, tabby-db/src/cache.rs
        let file_contents2 = include_str!("../../../../ee/tabby-db/src/cache.rs");

        let rust_chunks2 = CodeIntelligence::chunks(file_contents2, "rust")
            .map(|(_, chunk)| chunk)
            .collect::<Vec<_>>()
            .await;

        assert_snapshot!(format!("{:#?}", rust_chunks2));

        let text_chunks2 = CodeIntelligence::chunks(file_contents2, "unknown")
            .map(|(_, chunk)| chunk)
            .collect::<Vec<_>>()
            .await;

        assert_snapshot!(format!("{:#?}", text_chunks2));
    }
}
