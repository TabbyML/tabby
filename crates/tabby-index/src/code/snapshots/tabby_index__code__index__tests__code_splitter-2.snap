---
source: crates/tabby-index/src/code/index.rs
expression: "format!(\"{:#?}\", text_chunks)"
---
[
    "use dashmap::DashMap;\nuse tabby_common::languages::Language;\nuse trie_rs::{Trie, TrieBuilder};\n\npub struct StopConditionFactory {\n    stop_trie_cache: DashMap<String, Trie<u8>>,\n    stop_words_from_model_config: Vec<String>,\n}\n\nfn reverse<T>(s: T) -> String\nwhere\n    T: Into<String>,\n{\n    s.into().chars().rev().collect()\n}",
    "impl Default for StopConditionFactory {\n    fn default() -> Self {\n        Self {\n            stop_trie_cache: DashMap::new(),\n            stop_words_from_model_config: vec![],\n        }\n    }\n}\n\ntype CachedTrie<'a> = dashmap::mapref::one::Ref<'a, String, Trie<u8>>;\n\nimpl StopConditionFactory {\n    pub fn with_stop_words(stop_words: Vec<String>) -> Self {\n        Self {\n            stop_trie_cache: DashMap::new(),\n            stop_words_from_model_config: stop_words,\n        }\n    }",
    "pub fn create(&self, text: &str, language: Option<&'static Language>) -> StopCondition {\n        if let Some(language) = language {\n            StopCondition::new(self.get_trie(language), text)\n        } else {\n            StopCondition::new(None, text)\n        }\n    }",
    "fn get_trie<'a>(&'a self, language: &'static Language) -> Option<CachedTrie<'a>> {\n        let mut stop_words = language.get_stop_words();\n        // append model stop words\n        stop_words.extend(self.stop_words_from_model_config.iter().cloned());",
    "if stop_words.is_empty() {\n            None\n        } else {\n            let hashkey = language.language().to_owned();\n            let mut trie = self.stop_trie_cache.get(&hashkey);\n            if trie.is_none() {\n                self.stop_trie_cache\n                    .insert(hashkey.clone(), create_stop_trie(stop_words));\n                trie = self.stop_trie_cache.get(&hashkey);\n            }\n\n            trie\n        }\n    }\n}",
    "fn create_stop_trie(stop_words: Vec<String>) -> Trie<u8> {\n    let mut builder = TrieBuilder::new();\n    for word in stop_words {\n        builder.push(reverse(word))\n    }\n    builder.build()\n}\n\npub struct StopCondition<'a> {\n    stop_trie: Option<CachedTrie<'a>>,\n    reversed_text: String,\n    num_decoded: usize,\n}",
    "impl<'a> StopCondition<'a> {\n    pub fn new(stop_trie: Option<CachedTrie<'a>>, text: &str) -> Self {\n        Self {\n            stop_trie,\n            reversed_text: reverse(text),\n            num_decoded: 0,\n        }\n    }\n\n    pub fn should_stop(&mut self, new_text: &str) -> (bool, usize) {\n        self.num_decoded += 1;\n        if !new_text.is_empty() {\n            self.reversed_text = reverse(new_text) + &self.reversed_text;",
    "if let Some(re) = &self.stop_trie {\n                let matches = re.common_prefix_search(&self.reversed_text);\n                let matched_length = matches.into_iter().map(|x| x.len()).max();\n                if let Some(matched_length) = matched_length {\n                    return (true, matched_length);\n                }\n            }\n        }\n        (false, 0)\n    }\n}\n\n#[cfg(test)]\nmod tests {\n\n    use tabby_common::languages::UNKNOWN_LANGUAGE;\n\n    use super::*;",
    "#[test]\n    fn test_trie_works() {\n        let text = reverse(\"void write_u32(std::uint32_t val) const {\\n        write_raw(&val, sizeof(val));\\n    }\\n\\n    ~llama_file() {\\n        if (fp) {\\n            std::fclose(fp);\\n        }\\n    }\\n};\\n\\nvoid\");\n\n        let trie = create_stop_trie(vec![\"\\n\\n\".to_owned(), \"\\n\\n  \".to_owned()]);\n        assert!(trie.common_prefix_search(&text).is_empty());",
    "let trie = create_stop_trie(vec![\n            \"\\n\\n\".to_owned(),\n            \"\\n\\n  \".to_owned(),\n            \"\\nvoid\".to_owned(),\n            \"<|file_sep|>\".to_owned(), // qwen 2.5 coder style\n        ]);\n        assert!(!trie.common_prefix_search(&text).is_empty());\n\n        let qwen25coder = reverse(\"qwen25 style stop words;<|file_sep|>\");\n        assert!(!trie.common_prefix_search(&qwen25coder).is_empty());\n    }",
    "#[test]\n    fn test_stop_condition_max_length() {\n        let factory = StopConditionFactory::default();\n        let mut cond = factory.create(\"\", Some(&UNKNOWN_LANGUAGE));\n        let (should_stop, _) = cond.should_stop(\"1\");\n        assert!(!should_stop);\n        let (should_stop, _) = cond.should_stop(\"2\");\n        assert!(!should_stop);\n        let (should_stop, _) = cond.should_stop(\"3\");\n        assert!(!should_stop);\n        let (should_stop, _) = cond.should_stop(\"4\");\n        assert!(!should_stop)",
    "}\n\n    #[test]\n    fn test_stop_condition_additional_stop_words() {\n        let factory = StopConditionFactory::with_stop_words(vec![\"<|endoftext|>\".to_owned()]);\n        let mut cond = factory.create(\"\", Some(&UNKNOWN_LANGUAGE));\n        let (should_stop, _) = cond.should_stop(\"1\");\n        assert!(!should_stop);\n        let (should_stop, _) = cond.should_stop(\"<|endoftext|>\");\n        assert!(should_stop);\n    }\n}",
]
